nano/node/confirmation_height_processor.cpp

```diff
@@ -0,0 +1,426 @@
+	auto total_pending_write_block_count (total_pending_write_block_count_a);
+
+	// Write in batches
+	while (total_pending_write_block_count > 0)
+	{
+		uint64_t num_block_writes = 0;
+		auto transaction (store.tx_begin_write ());
+		while (!all_pending.empty ())
+		{
+			const auto & pending = all_pending.front ();
+			auto error = store.account_get (transaction, pending.account, account_info);
+			release_assert (!error);
+			if (pending.height > account_info.confirmation_height)
+			{
+#ifndef NDEBUG
+				// Do more thorough checking in Debug mode, indicates programming error.
+				nano::block_sideband sideband;
+				auto block = store.block_get (transaction, pending.hash, &sideband);
+				assert (block != nullptr);
+				assert (sideband.height == pending.height);
+#else
+				auto block = store.block_get (transaction, pending.hash);
+#endif
+				// Check that the block still exists as there may have been changes outside this processor.
+				if (!block)
+				{
+					logger.always_log ("Failed to write confirmation height for: ", pending.hash.to_string ());
+					ledger.stats.inc (nano::stat::type::confirmation_height, nano::stat::detail::invalid_block);
+					return true;
+				}
+
+				ledger.stats.add (nano::stat::type::confirmation_height, nano::stat::detail::blocks_confirmed, nano::stat::dir::in, pending.height - account_info.confirmation_height);
+				assert (pending.num_blocks_confirmed == pending.height - account_info.confirmation_height);
+				account_info.confirmation_height = pending.height;
+				store.account_put (transaction, pending.account, account_info);
+			}
+			total_pending_write_block_count -= pending.num_blocks_confirmed;
+			num_block_writes += pending.num_blocks_confirmed;
+			all_pending.erase (all_pending.begin ());
+
+			if (num_block_writes >= batch_write_size)
+			{
+				// Commit changes periodically to reduce time holding write locks for long chains
+				break;
+			}
+		}
+	}
+	return false;
+}
+
+/*
+ * This function assumes receive_source_pairs_lk is not already locked
+ */
+void nano::confirmation_height_processor::collect_unconfirmed_receive_and_sources_for_account (uint64_t block_height, uint64_t confirmation_height, nano::block_hash const & current, nano::account const & account, nano::transaction & transaction, std::unique_lock<std::mutex> & receive_source_pairs_lk)
+{
+	auto hash (current);
+	auto num_to_confirm = block_height - confirmation_height;
+
+	std::vector<receive_source_pair> temp_receive_source_pairs;
+
+	// Store heights of blocks
+	constexpr auto height_not_set = std::numeric_limits<uint64_t>::max ();
+	auto next_height = height_not_set;
+	while (num_to_confirm > 0 && !hash.is_zero ())
+	{
+		active.confirm_block (hash);
+		nano::block_sideband sideband;
+		auto block (store.block_get (transaction, hash, &sideband));
+		if (block)
+		{
+			auto source (block->source ());
+			if (source.is_zero ())
+			{
+				source = block->link ();
+			}
+
+			if (store.source_exists (transaction, source))
```

Comment:
I guess we can add check if source is_zero () or epoch link as well

Reply:
Good point, done.